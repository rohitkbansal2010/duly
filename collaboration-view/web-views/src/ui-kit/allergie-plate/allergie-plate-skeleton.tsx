import { DulyLoader } from '@components/duly-loader';
import { ALLERGIES_LOADER_WIDTH } from '@constants';

import styles from './allergie-plate.module.scss';

export const AllergiePlateSkeleton = () => 
  (
    <div className={styles.allergiePlateSkeleton}>
      <DulyLoader width={ALLERGIES_LOADER_WIDTH} />
    </div>
  );
