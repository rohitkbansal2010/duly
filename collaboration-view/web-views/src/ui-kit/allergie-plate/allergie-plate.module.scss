@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.allergiePlate {
  &Container {
    display: flex;
    flex-direction: column;
    height: 100%;
    padding: 1.25rem 1.25rem 1.5rem 1.25rem;
    background: white;
    border-radius: 0.5rem;
  }

  &TopPart {
    display: flex;
    align-items: center;
    margin-bottom: 1.75rem;

    &Icon {
      position: relative;
      flex-shrink: 0;
      box-sizing: border-box;
      width: 3.75rem;
      height: 3.75rem;
      padding: 0.75rem;
      background: lighten($dupage-light-blue, 40%);
      border-radius: 100%;

      &Img {
        width: 100%;
        height: 100%;
      }

      &Warning {
        position: absolute;
        top: -0.15rem;
        right: -0.5rem;
        width: 1.5rem;
        height: 1.5rem;
      }
    }

    &Title {
      @include nexa-heavy-highlight;

      margin-left: 1.25rem;
      font-size: var(--font-size-24px-4px);
      line-height: 1.625rem;
      word-break: break-word;
    }
  }

  &BottomPart {
    height: 100%;
    padding: 1.65rem 1.65rem 0.65rem;
    background: lighten($gray, 10%);
    border: 0.0625rem solid $gray;
    border-radius: 0.75rem;

    &Container {
      display: flex;
      align-items: flex-start;
      justify-content: space-between;
    }

    &Label {
      position: relative;
      display: flex;
      align-items: flex-start;
      margin-bottom: 0.25rem;
      color: $dark-gray;
      font-weight: 700;
      font-size: var(--font-size-14px-4px);
      line-height: var(--line-height-16px-4px);
      text-transform: uppercase;
    }

    &IconWarning {
      width: 1.2rem;
      height: 1.2rem;
      margin-left: 0.125rem;
      transform: translate(0, -20%);
      img {
        width: 100%;
        height: 100%;
      }
    }

    &Value {
      display: flex;
      flex-direction: column;
      font-weight: bold;
      font-size: var(--font-size-16px-4px);
      line-height: var(--line-height-24px-4px);

      &SeverityLevel {
        display: inline-block;
        width: 0.75rem;
        height: 0.75rem;
        margin-left: 0.5rem;
        border-radius: 100%;

        &--mild {
          background: $dupage-green;
        }

        &--moderate {
          background: $dupage-saffron;
        }

        &--severe {
          background: $dupage-orange;
        }
      }

      span {
        width: 90%;
        overflow-wrap: break-word;
      }
    }

    &Item {
      width: 25%;
    }

    &ReactionContainer {
      width: 50%;

      & > div[class^="allergiePlateBottomPartItem"] {
        width: 100%;
      }

      &Item {
        display: flex;
        width: 100%;

        span {
          width: 50%;
          padding-right: 5%;
          overflow-wrap: break-word;
        }
      }
    }
  }

  &Skeleton {
    position: relative;
    width: calc(50% - 0.625rem);
    margin-bottom: 1.25rem;
    background: $white;
    border-radius: 0.5rem;

    &:nth-last-child(-n + 2) {
      margin-bottom: 0;
    }
  }
}
