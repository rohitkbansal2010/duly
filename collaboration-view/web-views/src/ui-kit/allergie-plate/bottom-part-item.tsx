import { PropsWithChildren, ReactNode } from 'react';

import styles from './allergie-plate.module.scss';

type BottomPartItemProps = PropsWithChildren<{ title: ReactNode }>;

export const BottomPartItem = ({ title, children }: BottomPartItemProps) => 
  (
    <div className={styles.allergiePlateBottomPartItem}>
      <div className={styles.allergiePlateBottomPartLabel}>{title}</div>
      <div className={styles.allergiePlateBottomPartValue}>{children}</div>
    </div>
  );
