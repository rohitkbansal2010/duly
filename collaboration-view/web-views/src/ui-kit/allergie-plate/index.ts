export * from './allergie-plate';
export * from './allergie-plate-skeleton';
