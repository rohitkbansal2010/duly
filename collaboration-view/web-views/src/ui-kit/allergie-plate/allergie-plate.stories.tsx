import { AllergiePlate } from './allergie-plate';

export default {
  title: 'UI-KIT / AllergiePlate',
  component: AllergiePlate,
};

export const Example = () => 
  (
    <div style={{ background: '#ddd', width: '600px', padding: 40 }}>
      <AllergiePlate id="1" title="Peanuts" recorded="11/2/20" categories={[ 'food' ]} reactions={[ { title: 'anaphylaxis', severity: 'severe' } ]} />
    </div>
  );
