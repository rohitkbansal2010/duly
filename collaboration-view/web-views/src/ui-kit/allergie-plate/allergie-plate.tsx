import capitalize from 'lodash/capitalize';
import sortBy from 'lodash/sortBy';
import { useMemo } from 'react';

import {
  environmentAllergieIcon,
  foodAllergieIcon,
  biologicAllergieIcon,
  medicalAllergieIcon,
  severeLevelWarning,
  noAllergiesIcon as commonAllergyIcon, otherAllergiesIcon
} from '@icons';
import { AllergieCategoryType, AllergiesData } from '@types';
import { formatMDYYYYDate } from '@utils';

import styles from './allergie-plate.module.scss';
import { BottomPartItem } from './bottom-part-item';

const IconsMap: {
  [K in AllergieCategoryType]: string;
} = {
  Food: foodAllergieIcon,
  Medication: medicalAllergieIcon,
  Biologic: biologicAllergieIcon,
  Environment: environmentAllergieIcon,
  Other: otherAllergiesIcon,
};

const getIcon = (categories: AllergieCategoryType[] = []): string => {
  if (categories.length > 1 || categories.length === 0) {
    return commonAllergyIcon;
  }
  return IconsMap[categories[0]];
};

type SeverityItemLabelProps = { isSevereAllergie: boolean };

const SeverityItemLabel = ({ isSevereAllergie }: SeverityItemLabelProps) =>
  (
    <>
      Severity
      {isSevereAllergie && (
      <img
        className={styles.allergiePlateBottomPartIconWarning}
        src={severeLevelWarning}
        alt="severe level warning"
      />
      )}
    </>
  );

export const AllergiePlate = ({
  title = '',
  recorded = '',
  categories = [],
  reactions = [],
}: AllergiesData) => {
  const iconSrc = getIcon(categories);

  const sortedCategories = useMemo(() =>
    categories.sort(), [ categories ]);

  const isSevereAllergy = useMemo(
    () =>
      reactions.some(({ severity }) =>
        severity === 'Severe'),
    [ reactions ]
  );

  const sortedReactions = useMemo(() =>
    sortBy(reactions, [ 'title' ]), [ reactions ]);

  return (
    <div
      className={styles.allergiePlateContainer}
      data-testid="allergie-plate-item"
    >
      <div className={styles.allergiePlateTopPart}>
        <div className={styles.allergiePlateTopPartIcon}>
          <img className={styles.allergiePlateTopPartIconImg} src={iconSrc} alt="allergie icon" />
          {isSevereAllergy && (
            <img
              className={styles.allergiePlateTopPartIconWarning}
              src={severeLevelWarning}
              alt="severe level warning"
            />
          )}
        </div>
        <div className={styles.allergiePlateTopPartTitle}>{capitalize(title)}</div>
      </div>
      <div className={styles.allergiePlateBottomPart}>
        <div className={styles.allergiePlateBottomPartContainer}>
          <BottomPartItem title={<span data-testid="allergie-plate-date-recorded-title">Date Recorded</span>}>
            <span data-testid="allergie-plate-date-recorded-value">{formatMDYYYYDate(recorded)}</span>
          </BottomPartItem>
          <BottomPartItem title={<span data-testid="allergie-plate-category-title">Category</span>}>
            {sortedCategories.map(category =>
            (
              <span data-testid="allergie-plate-category-item" data-testval={capitalize(category)} key={category}>{capitalize(category)}</span>
            ))}
          </BottomPartItem>
          <div className={styles.allergiePlateBottomPartReactionContainer}>
            <BottomPartItem
              title={(
                <div className={styles.allergiePlateBottomPartReactionContainerItem}>
                  <span data-testid="allergie-plate-severity-title">
                    <SeverityItemLabel isSevereAllergie={isSevereAllergy} />
                  </span>
                  <span data-testid="allergie-plate-reaction-title">Reaction</span>
                </div>
              )}
            >
              {sortedReactions.map(({ title, severity }) =>
              (
                <div
                  key={title + severity}
                  className={styles.allergiePlateBottomPartReactionContainerItem}
                >
                  <span data-testid="allergie-plate-severity-item" data-testval={capitalize(severity)}>{capitalize(severity)}</span>
                  <span data-testid="allergie-plate-reaction-item" data-testval={capitalize(title)}>{capitalize(title)}</span>
                </div>
              ))}
            </BottomPartItem>
          </div>
        </div>
      </div>
    </div>
  );
};
