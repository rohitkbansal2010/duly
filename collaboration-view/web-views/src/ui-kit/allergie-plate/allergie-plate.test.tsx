import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import capitalize from 'lodash/capitalize';

import {
  foodAllergieIcon,
  medicalAllergieIcon,
  biologicAllergieIcon,
  environmentAllergieIcon,
  noAllergiesIcon,
  severeLevelWarning
} from '@icons';
import { allergiesMockData } from '@mock-data';
import { formatMDYYYYDate } from '@utils';

import { AllergiePlate } from './allergie-plate';
import { AllergiePlateSkeleton } from './allergie-plate-skeleton';

describe('AllergiesPlate', () => {
  let component: ReturnType<typeof mount | typeof shallow>;
  const mockProps = allergiesMockData[0];

  it('component should match snapshot', () => {
    component = shallow(<AllergiePlate {...mockProps} />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('skeleton should match snapshot', () => {
    component = shallow(<AllergiePlateSkeleton />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render right data', () => {
    component = mount(<AllergiePlate {...mockProps} />);

    const {
      title,
      categories,
      reactions: [ { title: reactionsTitle } ],
      recorded,
    } = mockProps;

    expect(component.text().includes(capitalize(title!))).toBeTruthy();
    expect(component.text().includes(capitalize(reactionsTitle))).toBeTruthy();
    expect(component.text().includes(categories.join(''))).toBeTruthy();
    expect(component.text().includes(formatMDYYYYDate(recorded))).toBeTruthy();
  });

  it('should render severity icon', () => {
    component = shallow(<AllergiePlate {...mockProps} />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === severeLevelWarning)).toHaveLength(1);
  });

  it('should not render severity icon', () => {
    component = shallow(<AllergiePlate {...allergiesMockData[1]} />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === severeLevelWarning)).toHaveLength(0);
  });

  it('should render food category icon', () => {
    component = shallow(<AllergiePlate {...{ ...mockProps, categories: [ 'Food' ] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === foodAllergieIcon)).toHaveLength(1);
  });

  it('should render medication category icon', () => {
    component = shallow(<AllergiePlate {...{ ...mockProps, categories: [ 'Medication' ] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === medicalAllergieIcon)).toHaveLength(1);
  });

  it('should render biologic category icon', () => {
    component = shallow(<AllergiePlate {...{ ...mockProps, categories: [ 'Biologic' ] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === biologicAllergieIcon)).toHaveLength(1);
  });

  it('should render environment category icon', () => {
    component = shallow(<AllergiePlate {...{ ...mockProps, categories: [ 'Environment' ] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === environmentAllergieIcon)).toHaveLength(1);
  });

  it('should render common icon', () => {
    component = shallow(<AllergiePlate {...{ ...mockProps, categories: [ 'Environment', 'Biologic' ] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === noAllergiesIcon)).toHaveLength(1);

    component = shallow(<AllergiePlate { ...{ ...mockProps, categories: [] } } />);

    expect(component.findWhere(n => 
      n.type() === 'img' && n.props().src === noAllergiesIcon)).toHaveLength(1);
  });
});
