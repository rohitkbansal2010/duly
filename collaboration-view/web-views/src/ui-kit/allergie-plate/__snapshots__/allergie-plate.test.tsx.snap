// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllergiesPlate component should match snapshot 1`] = `
<div
  className="allergiePlateContainer"
  data-testid="allergie-plate-item"
>
  <div
    className="allergiePlateTopPart"
  >
    <div
      className="allergiePlateTopPartIcon"
    >
      <img
        alt="allergie icon"
        className="allergiePlateTopPartIconImg"
        src="food-allergie-icon.svg"
      />
      <img
        alt="severe level warning"
        className="allergiePlateTopPartIconWarning"
        src="severe-level-warning.svg"
      />
    </div>
    <div
      className="allergiePlateTopPartTitle"
    >
      Peanuts
    </div>
  </div>
  <div
    className="allergiePlateBottomPart"
  >
    <div
      className="allergiePlateBottomPartContainer"
    >
      <BottomPartItem
        title={
          <span
            data-testid="allergie-plate-date-recorded-title"
          >
            Date Recorded
          </span>
        }
      >
        <span
          data-testid="allergie-plate-date-recorded-value"
        >
          2/21/2022
        </span>
      </BottomPartItem>
      <BottomPartItem
        title={
          <span
            data-testid="allergie-plate-category-title"
          >
            Category
          </span>
        }
      >
        <span
          data-testid="allergie-plate-category-item"
          data-testval="Food"
          key="Food"
        >
          Food
        </span>
      </BottomPartItem>
      <div
        className="allergiePlateBottomPartReactionContainer"
      >
        <BottomPartItem
          title={
            <div
              className="allergiePlateBottomPartReactionContainerItem"
            >
              <span
                data-testid="allergie-plate-severity-title"
              >
                <SeverityItemLabel
                  isSevereAllergie={true}
                />
              </span>
              <span
                data-testid="allergie-plate-reaction-title"
              >
                Reaction
              </span>
            </div>
          }
        >
          <div
            className="allergiePlateBottomPartReactionContainerItem"
            key="AnaphylaxisSevere"
          >
            <span
              data-testid="allergie-plate-severity-item"
              data-testval="Severe"
            >
              Severe
            </span>
            <span
              data-testid="allergie-plate-reaction-item"
              data-testval="Anaphylaxis"
            >
              Anaphylaxis
            </span>
          </div>
        </BottomPartItem>
      </div>
    </div>
  </div>
</div>
`;

exports[`AllergiesPlate skeleton should match snapshot 1`] = `
<div
  className="allergiePlateSkeleton"
>
  <DulyLoader
    width={8}
  />
</div>
`;
